crates/astria-composer/src/searcher/collector.rs

```diff
@@ -0,0 +1,221 @@
+use std::time::Duration;
+
+use color_eyre::eyre::{
+    self,
+    WrapErr as _,
+};
+use ethers::providers::{
+    Provider,
+    ProviderError,
+    Ws,
+};
+use humantime::format_duration;
+use tokio::sync::{
+    mpsc::{
+        error::SendTimeoutError,
+        Sender,
+    },
+    watch,
+};
+use tracing::{
+    debug,
+    info,
+    instrument,
+    warn,
+};
+
+/// A wrapper around an [`ethers::types::Transaction`] that includes the chain ID.
+///
+/// Used to send new transactions to the searcher.
+pub(super) struct Transaction {
+    pub(super) chain_id: String,
+    pub(super) inner: ethers::types::Transaction,
+}
+
+#[derive(Debug)]
+pub(super) struct Collector {
+    // Chain ID to identify in the astria sequencer block which rollup a serialized sequencer
+    // action belongs to.
+    chain_id: String,
+    // The client for getting new pending transactions from an ethereum rollup.
+    client: EthClient,
+    // The channel on which the collector sends new txs to the searcher.
+    searcher_channel: Sender<Transaction>,
+    // The status of this collector.
+    status: watch::Sender<Status>,
+}
+
+#[derive(Debug)]
+pub(super) struct Status {
+    is_connected: bool,
+}
+
+impl Status {
+    fn new() -> Self {
+        Self {
+            is_connected: false,
+        }
+    }
+
+    pub(super) fn is_connected(&self) -> bool {
+        self.is_connected
+    }
+}
+
+impl Collector {
+    pub(super) async fn new(
+        chain_id: String,
+        url: String,
```

Comment:
Imo, the collector should accept a client instead of initializing it (Single responsibility)

Reply:
I disagree - a `Collector` should be responsible for initializing a client, pulling txs from it and forwarding them for bundling. I don't see why we need an additional abstraction layer between the client and the `Collector`. If we want to use a different client then there should be a different `Collector` implementation around it.

For working with multiple rollups we should have multiple Collectors.

Reply:
Personally, I would add a layer of abstraction so that collectors being generic is a simple change in the future. Plus, it's not an abstraction at all at this point, you are just initializing somewhere else. Eventually, the collector will need to accept a generic type of a node that it can connect to, which I believe should be a trait. But it's more of a design decision at the end of the day

Reply:
Different collector implementation per client doesn't make sense to me because why would we want people to copy paste wiring code?

Reply:
I think it's a bit early to address this, and making this change before we have a different client to implement this for adds unneeded complexity. I think that when we deal with a different rollup node we're inevitably going to have to change whatever abstraction we made anyways, so no need to do this here.

Reply:
This is out of scope. This PR is only concerned with reading from different rollups that serve the same RPC.

We should think about generalization once we need it (for example once we create a crate to implement bespoke composers).